### PR TITLE
[HttpClient] add ResponseSetMonitor to trigger callbacks on response completion

### LINKED ISSUE #32142: [HttpClient] Getting responses asynchronously as soon as they ready
Here is the deal: I use Symfony and HttpClient in telegram bot. Bot sends various requests to Telegram's API, and most of these I don't really care to get the response in main bot logic (I want to run the code as fast as possible). However, I'd like to add requests and responses to Sentry breadcrumbs for reporting. Adding requests is easy. Adding responses is tricky.

I don't want actually to call `Response->getContent()`, so that it would stop main logic execution until I get the response. However, I'd like to somehow get notified as soon as response is ready. Then, I'd like to add the response to breadcrumbs, and then if there were errors in responses - send a notification.

I've tried subscribing to `on_progress`, but that didn't work, since responses were not ready when I tried using them from that callback.

So, I couldn't find anything about some sort of event system, so that I could subscribe and get responses _as soon as they get ready_. Is it really possible? And if not, then it's a feature request, I guess...

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | Fix #32142
| License       | MIT
| Doc PR        | -

Related to #32367

Follows some discussions on Slack+Github: how do we trigger callbacks when responses complete?
Here is a quick poc doing so:
```php
$client = HttpClient::create();
$monitor = new ResponseSetMonitor($client);
$monitor->add($client->request(...), $onHeaders, $onBody, $onError);
// [...]
$monitor->tick(); // non-blocking call
// [...]
$monitor->tick(0.001); // don't make the tick last more than 1ms
// [...]
$monitor->complete(); // complete responses and consider timeouts as errors
// [...]
$monitor->cancel() // cancel all pending responses - usually run after a ->tick($someTimeout)
```

Of course, since there is no event loop under the hood, one still needs to "tick" to move the network read/writes forward.

Useful in the component? If you think so I'd happily accept a PR that adds tests on top :)

### CODE REVIEW

File: src/Symfony/Component/HttpClient/ResponseSetMonitor.php [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        $this->responses = new \SplObjectStorage();
+    }
+
+    public function add(ResponseInterface $response, callable $onHeaders = null, callable $onBody = null, callable $onError = null): void
+    {
+        $this->responses[$response] = [$onHeaders, $onBody, $onError];
+        $this->tick();
+    }
+
+    public function tick(?float $idleTimeout = 0.0): int
```
Comment (lyrixx) [inline src/Symfony/Component/HttpClient/ResponseSetMonitor.php]: I read the code in master, and passing `null` means `INF`. IMHO there is no need to all passing explicitly `null` here.
I would remove `?` in the typehint.

Comment (nicolas-grekas) [reply src/Symfony/Component/HttpClient/ResponseSetMonitor.php]: `null` actually means to use the default request's timeout.
But I still removed the `?` and I changed the meaning of the timeout: now it's a "max wait time", not only a "max idle time".

File: src/Symfony/Component/HttpClient/ResponseSetMonitor.php [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+ */
+final class ResponseSetMonitor implements \Countable
+{
+    private $client;
+    private $onHeaders;
+    private $onBody;
+    private $onError;
+    private $responses;
+
+    public function __construct(HttpClientInterface $client, callable $onHeaders = null, callable $onBody = null, callable $onError = null)
```
Comment (lyrixx) [inline src/Symfony/Component/HttpClient/ResponseSetMonitor.php]: IMHO, the onHeaders callable should be the last argument, it's less useful than the body or error.
Same for `add` method

File: src/Symfony/Component/HttpClient/ResponseSetMonitor.php [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+                if (null !== $onHeaders && $chunk->isFirst()) {
+                    $onHeaders($response);
+                }
+
+                if (null !== $onBody && $chunk->isLast()) {
+                    $onBody($response);
+                }
+
+                if (null === $onBody || $chunk->isLast()) {
```
Comment (lyrixx) [inline src/Symfony/Component/HttpClient/ResponseSetMonitor.php]: why checking `null === onBody` ? if it's the last chunk, it should be removed from the responses stack.
IMHO, there is a leak here

File: src/Symfony/Component/HttpClient/ResponseSetMonitor.php [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+                    unset($this->responses[$response]);
+                }
+            } catch (ExceptionInterface $e) {
+                [, , $onError] = $this->responses[$response];
+                $onError = $onError ?? $this->onError;
+                unset($this->responses[$response]);
+
+                if (null !== $onError) {
```
Comment (lyrixx) [inline src/Symfony/Component/HttpClient/ResponseSetMonitor.php]: if would use a positive if:
```php
if($onError) {
```

File: src/Symfony/Component/HttpClient/ResponseSetMonitor.php [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+                    $error = $error ?? $e;
+                }
+            } finally {
+                if (!$errorOnTimeout && $remainingTimeout) {
+                    $remainingTimeout = max(0.0, $timeout - microtime(true) + $startTime);
+                }
+            }
+        }
+
+        if (null !== $error) {
```
Comment (lyrixx) [inline src/Symfony/Component/HttpClient/ResponseSetMonitor.php]: Only the last failed response will throw an error

File: src/Symfony/Component/HttpClient/ResponseSetMonitor.php [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+        $this->wait(null, true);
+    }
+
+    private function wait(?float $timeout, bool $errorOnTimeout): int
+    {
+        $error = null;
+        $remainingTimeout = $timeout;
+
+        if (!$errorOnTimeout && $remainingTimeout) {
+            $startTime = microtime(true);
```
Comment (lyrixx) [inline src/Symfony/Component/HttpClient/ResponseSetMonitor.php]: Why using a system why startTime when you can decrease remainingTimeout and check when it's <0 ?

File: src/Symfony/Component/HttpClient/ResponseSetMonitor.php [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+        }
+
+        foreach ($this->client->stream($this->responses, $remainingTimeout) as $response => $chunk) {
+            try {
+                if ($chunk->isTimeout() && !$errorOnTimeout) {
+                    continue;
```
Comment (lyrixx) [inline src/Symfony/Component/HttpClient/ResponseSetMonitor.php]: The response should be removed from the responses stack

Comment (lyrixx) [reply src/Symfony/Component/HttpClient/ResponseSetMonitor.php]: Okay, got it. The responses is already removed. 
Sometimes it is removed, sometimes it it not ¯\_(ツ)_/¯
And all stack trace I got are really hard to debug:
```
PHP Fatal error:  Uncaught Symfony\Component\HttpClient\Exception\TransportException: Couldn't resolve host name for "http://klmfdslmkfksd.com/robots.txt". in /home/rio/rio/server/backend/vendor/symfony/symfony/src/Symfony/Component/HttpClient/Response/CurlResponse.php:264
Stack trace:
#0 /home/rio/rio/server/backend/vendor/symfony/symfony/src/Symfony/Component/HttpClient/Response/ResponseTrait.php(276): Symfony\Component\HttpClient\Response\CurlResponse::perform(Object(Symfony\Component\HttpClient\Internal\CurlClientState), Array)
#1 [internal function]: Symfony\Component\HttpClient\Response\CurlResponse::stream(Array)
#2 /home/rio/rio/server/backend/vendor/symfony/symfony/src/Symfony/Component/HttpClient/Response/CurlResponse.php(121): Generator->current()
#3 /home/rio/rio/server/backend/vendor/symfony/symfony/src/Symfony/Component/HttpClient/Response/ResponseTrait.php(67): Symfony\Component\HttpClient\Response\CurlResponse::Symfony\Component\HttpClient\Response\{closure}(Object(Symfony\Component\HttpClient\Response\CurlResp in /home/rio/rio/server/backend/vendor/symfony/symfony/src/Symfony/Component/HttpClient/Chunk/ErrorChunk.php on line 105
```
As you can see, I did not see **my** code. :cry:

**Comment by nicolas-grekas** [comment]
#33743 will provide a better abstraction and #32367 is here to keep track of a lower level / native implementation.

**Pull request closed without merging**